src/util-landlock.c

```diff
@@ -0,0 +1,269 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Eric Leblond <el@stamus-networks.com>
+ */
+
+#include "suricata.h"
+#include "util-conf.h"
+#include "util-landlock.h"
+#include "util-mem.h"
+
+#ifndef HAVE_LINUX_LANDLOCK_H
+
+void LandlockSandboxing(SCInstance *suri)
+{
+    return;
+}
+
+#else /* HAVE_LINUX_LANDLOCK_H */
+
+#include <linux/landlock.h>
+
+#ifndef landlock_create_ruleset
+static inline int landlock_create_ruleset(
+        const struct landlock_ruleset_attr *const attr, const size_t size, const __u32 flags)
+{
+    return syscall(__NR_landlock_create_ruleset, attr, size, flags);
+}
+#endif
+
+#ifndef landlock_add_rule
+static inline int landlock_add_rule(const int ruleset_fd, const enum landlock_rule_type rule_type,
+        const void *const rule_attr, const __u32 flags)
+{
+    return syscall(__NR_landlock_add_rule, ruleset_fd, rule_type, rule_attr, flags);
+}
+#endif
+
+#ifndef landlock_restrict_self
+static inline int landlock_restrict_self(const int ruleset_fd, const __u32 flags)
+{
+    return syscall(__NR_landlock_restrict_self, ruleset_fd, flags);
+}
+#endif
+
+#ifndef LANDLOCK_ACCESS_FS_REFER
+#define LANDLOCK_ACCESS_FS_REFER (1ULL << 13)
+#endif
+
+#define _LANDLOCK_ACCESS_FS_WRITE                                                                  \
+    (LANDLOCK_ACCESS_FS_WRITE_FILE | LANDLOCK_ACCESS_FS_REMOVE_DIR |                               \
+            LANDLOCK_ACCESS_FS_REMOVE_FILE | LANDLOCK_ACCESS_FS_MAKE_CHAR |                        \
+            LANDLOCK_ACCESS_FS_MAKE_DIR | LANDLOCK_ACCESS_FS_MAKE_REG |                            \
+            LANDLOCK_ACCESS_FS_MAKE_SOCK | LANDLOCK_ACCESS_FS_MAKE_FIFO |                          \
+            LANDLOCK_ACCESS_FS_MAKE_BLOCK | LANDLOCK_ACCESS_FS_MAKE_SYM |                          \
+            LANDLOCK_ACCESS_FS_REFER)
+
+#define _LANDLOCK_ACCESS_FS_READ (LANDLOCK_ACCESS_FS_READ_FILE | LANDLOCK_ACCESS_FS_READ_DIR)
+
+#define _LANDLOCK_SURI_ACCESS_FS_WRITE                                                             \
+    (LANDLOCK_ACCESS_FS_WRITE_FILE | LANDLOCK_ACCESS_FS_MAKE_DIR | LANDLOCK_ACCESS_FS_MAKE_REG |   \
+            LANDLOCK_ACCESS_FS_REMOVE_FILE | LANDLOCK_ACCESS_FS_MAKE_SOCK)
+
+struct landlock_ruleset {
+    int fd;
+    struct landlock_ruleset_attr attr;
+};
+
+static inline struct landlock_ruleset* LandlockCreateRuleset()
+{
+    struct landlock_ruleset *ruleset = SCCalloc(1, sizeof(struct landlock_ruleset));
+    if (ruleset == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Can't alloc landlock ruleset");
+        return NULL;
+    }
+
+    ruleset->attr.handled_access_fs =
+                _LANDLOCK_ACCESS_FS_READ | _LANDLOCK_ACCESS_FS_WRITE | LANDLOCK_ACCESS_FS_EXECUTE;
+
+    int abi = landlock_create_ruleset(NULL, 0, LANDLOCK_CREATE_RULESET_VERSION);
+    if (abi < 0) {
+        SCFree(ruleset);
+        return NULL;
+    }
+    if (abi < 2) {
```

Comment:
I think the same comment still applies? (Missed to add the change before pushing?)

https://github.com/OISF/suricata/pull/7696#discussion_r938885139

Reply:
Suricata is only doing rename if file store (extraction of file from flow) is enabled so I'm checking this in the next line. Did I miss something ?

Reply:
If the kernel only supports Landlock ABI v1, with the current code

```
ruleset->attr.handled_access_fs &= ~LANDLOCK_ACCESS_FS_REFER;
```

the `rename()` operations will not work between directories on such kernels. (before kernel v5.19, these `rename()` attempts will result in EXDEV, as if the directories were part of different file systems).

Reply:
The only Suricata feature that is using rename is the file store. So the code is checking that this is not enable (when abi is 1) and it disables the Landlock.

Reply:
It does sound like a misunderstanding to me.

Renaming (moving) files between directories does **not** work when a Landlock V1 policy is enforced, even when the "refer" flag is not set. Removing the "refer" flag from handled_access_fs (as done on line 103) does not make it possible to move files between directories.

This is documented in https://docs.kernel.org/userspace-api/landlock.html#file-renaming-and-linking-abi-1

(I could not find a place here in the code where it is distinguishing between the case where "file store" is needed and where it isn't, so I assume that you are referring to line 103?)

Reply:
The final code is correct but the **GitHub interface is very confusing** because I guess both @gnoack and @regit are reviewing a different commit: https://github.com/OISF/suricata/pull/7697/commits/70d737658c2ed1bccfc52c7a14dc01facad09e78 (which only checks the ABI) and https://github.com/OISF/suricata/pull/7697/commits/f3ff53547dca8857800d1d9e0765935ae2c722bd (which adds the `RequiresFeature(FEATURE_OUTPUT_FILESTORE)` check), but the same conversation is visible in each page, then with different code :clown_face:… So, I think you're both right, but your "next line" (103 or 105) is relative, and the [final code](https://github.com/OISF/suricata/pull/7697/files) is good! :upside_down_face:

Reply:
Oh good point! 💡 I mixed this up completely and I retract the previous comment! The new code looks correct. 👍